### ISSUE TITLE
[BUG] AKS 1.25 removal of failure-domain.beta.kubernetes.io/region label breaks PV mounts

### ISSUE THREAD

**Issue by awangptc** [Azure/AKS#3668]
This is the same problem as
https://github.com/Azure/AKS/issues/3563

But the previous issue was locked with an answer provided of
```
the pv label actually does not matter, while the nodeAffinity field matters, since you are using in-tree azure disk volume, it's a good time migrating to csi driver volume, and in the meantime, change failure-domain.beta.kubernetes.io/region to topology.kubernetes.io/region, follow by: https://learn.microsoft.com/en-us/azure/aks/csi-migrate-in-tree-volumes
```

I believe this is an inappropriate and incomplete response.

We have numerous PVs that use the csi StorageClass.  At some point the csi-driver was changed to create a new nodeAffinity for the PV, but legacy - pre-existing PVs still use the old nodeAffinity.  This is not just a matter of migrating from in-tree to csi.  This occurs with CSI volumes.

It feels inappropriate to require migrating these CSI PVs to a new CSI PV.

**Comment by andyzhangx** [comment]
hi @awangptc actually AKS 1.26 removes `failure-domain.beta.kubernetes.io/region` and `failure-domain.beta.kubernetes.io/zone` label, and AKS 1.25 still works with those two deprecated labels, and that's a decision made from k8s community. Those labels are all created by in-tree azure disk driver which is already deprecated, and AKS won't support those in-tree volumes from 1.26.

The nodeAffinity property of a persistent volume is immutable, so the only way to change it is to recreate the PV and PVC. Fortunately, AKS provides an automated script to migrate in-tree PVs to CSI PVs with correct labels. You can find the script and instructions on how to use it in the following link:  https://learn.microsoft.com/en-us/azure/aks/csi-migrate-in-tree-volumes#migrate-disk-volumes

**Comment by andyzhangx** [comment]
> If the PVC is managed by a helm deployment and the helm chart isn't capable of allowing you to manipulate it in a way that would accept the migrated resources, isn't this a problem? It's a deficiency with the helm chart to not allow this, but at the same time, this is not something that the helm chart author's may even be aware is a necessary thing.

@awangptc I think your helm deployment was still using in-tree driver storage class at that time?
 btw, currently manual label a node with `failure-domain.beta.kubernetes.io/region` is prohibited (check below error msg), we are going to remove a restriction from 05-21 AKS release, by that way, user could workaround this issue temporarily, while in long term, we suggest migrating to csi driver volume with new affinity value.
```
# kubectl label node aks-agentpool-20657377-vmss000002 failure-domain.beta.kubernetes.io/region=eastus2
Error from server: admission webhook "aks-node-validating-webhook.azmk8s.io" denied the request: (UID: a8f77368-2c7c-44ca-b723-015cd817f9d4) User is trying to add or update a non-changeable system label (failure-domain.beta.kubernetes.io/region:eastus2). This action is denied..
```

**Comment by miwithro** [comment]
The announcement of the breaking change on the AKS side was in our latest AKS release notes https://github.com/Azure/AKS/releases/tag/2023-05-28

**Comment by jimmy03190** [comment]
Hi all, I'm having the same issue when upgrading to aks 1.26.3. I've read the whole article(https://learn.microsoft.com/en-us/azure/aks/csi-migrate-in-tree-volumes#migrate-disk-volumes) and still wondering how this thing should be handle when you only manage prometheus-stack with helm. I couldn't find any related issue on prometheus or helm community so I just post it here. Did anyone find a solution (working with helm deployment plus using deprecated lable) for this issue?

**Comment by andyzhangx** [comment]
> Hi all, I'm having the same issue when upgrading to aks 1.26.3. I've read the whole article(https://learn.microsoft.com/en-us/azure/aks/csi-migrate-in-tree-volumes#migrate-disk-volumes) and still wondering how this thing should be handle when you only manage prometheus-stack with helm. I couldn't find any related issue on prometheus or helm community so I just post it here. Did anyone find a solution (working with helm deployment plus using deprecated lable) for this issue?

@jimmy03190 those deprecated labels are generated with in-tree azure disk driver on AKS version < 1.21, and we won't have such issue if disk volume are created on AKS version >= 1.21, and we will rollout a fix that could regenerate the deprecated labels on AKS 1.26 in 0618 release(not rolled out yet).

**Comment by andyzhangx** [comment]
> @andyzhangx Thanks for the fast reply. Few things I would like to clarify: (1) The fix could only generate deprecated labels, and users still needs to manually fix each label by themselves? (2) If I re-created my disk volume with official suggestion(https://learn.microsoft.com/en-us/azure/aks/csi-migrate-in-tree-volumes#migrate-disk-volumes). The issue still remains (How do you update new PVC to existing app/pod if your app is managed by HELM chart?)

@jimmy03190 the fix could regenerate deprecated labels on the node, so the existing pv with deprecated label affinity could still work, no manual action needed with that fix.
about the migration, you could create new PVC bound to new PV with new labels, and the replace existing deployments with new created PVCs.

if you can't wait for AKS 0618 release, we could do a quick migration in our backend if you could provide the cluster api-server address through azure support.

**Comment by jimmy03190** [comment]
@andyzhangx Thanks for quick response.
So you're saying AKS 0618 release will regenerate with new labels on the node, so PV with in-tree format could still work? or not?
Does that mean if I replace my node with "topology.kubernetes.io" then the AKS Upgrade can still be proceed? Or the issue  still remains due to "in-tree" azure disk?

**Comment by andyzhangx** [comment]
> @andyzhangx Thanks for quick response. So you're saying AKS 0618 release will regenerate with new labels on the node, so PV with in-tree format could still work? or not? Does that mean if I replace my node with "topology.kubernetes.io" then the AKS Upgrade can still be proceed? Or the issue still remains due to "in-tree" azure disk?

@jimmy03190 AKS 0618 release will regenerate deprecated and new labels on the node, so PV with in-tree format could still work.

**Comment by andyzhangx** [comment]
> @andyzhangx Thanks for providing an update to this. I was in contact with the support team last week and they stated the following:
> 
> "we would like to bring to your attention that, until Kubernetes 1.28 (expected in October 2023), AKS will continue to generate deprecated labels on the node"
> 
> So just to confirm/clarify; The regeneration of the labels starting with the 0618 release is only a temporary fix until Kubernetes 1.28. This is only supposed to give customers more time to migrate the old in-tree volumes, right?

@mdnix that's correct, we won't generate the deprecated labels on the node forever, and old in-tree volume is not officially supported (though it still works) from AKS 1.26

**Comment by jimmy03190** [comment]
@andyzhangx Hi Andy, just want to confirm AKS 0618 patch will be released shortly since I've check AKS release note (https://github.com/Azure/AKS/releases) and couldn't find any related information. Please kindly advise. Thank you.

@mdnix This solution point to new PV right after creation. So there might be "down time" depending on size of PV? Since the script still backup origin data in old PV.

**Comment by andyzhangx** [comment]
> @andyzhangx Hi Andy, just want to confirm AKS 0618 patch will be released shortly since I've check AKS release note (https://github.com/Azure/AKS/releases) and couldn't find any related information. Please kindly advise. Thank you.
> 
> @mdnix This solution point to new PV right after creation. So there might be "down time" depending on size of PV? Since the script still backup origin data in old PV.

@jimmy03190 yes, 0618 release would begin to rollout this week.

**Comment by jimmy03190** [comment]
@andyzhangx So I guess UAE North still needs to wait for rollout of 0618 release?
https://releases.aks.azure.com

**Comment by mdnix** [comment]
> @andyzhangx Hi Andy, just want to confirm AKS 0618 patch will be released shortly since I've check AKS release note (https://github.com/Azure/AKS/releases) and couldn't find any related information. Please kindly advise. Thank you.
> 
> @mdnix This solution point to new PV right after creation. So there might be "down time" depending on size of PV? Since the script still backup origin data in old PV.

@jimmy03190 The recreation of the PV and  PVC might cause a brief period of downtime, as it involves temporary scaling down of the Pods. However, despite this, the size of the disk in Azure is not expected to impact the duration of this downtime. This is because the disk already exists, ensuring the preservation of original data and continuity of Pod functionality as before.

**Comment by andyzhangx** [comment]
correction: 0625 release would regenerate the deprecated labels

**Comment by jimmy03190** [comment]
@andyzhangx You mean 0618 release still remain the same error if we don't manually remove "old labels"? Is there an ETA for 0625 release? Cause I didn't see any release from (https://releases.aks.azure.com/)

**Comment by andyzhangx** [comment]
> @andyzhangx You mean 0618 release still remain the same error if we don't manually remove "old labels"? Is there an ETA for 0625 release? Cause I didn't see any release from (https://releases.aks.azure.com/)

correct, and 0625 release is already on canary region(eastus2euap), it still needs around 10 days to rollout complete on all regions.

**Comment by jimmy03190** [comment]
Hi @andyzhangx , I've replaced deprecated label with "topology" and try to upgrade AKS 1.26. However, I still encounter the same issue as this thread(https://github.com/Azure/AKS/issues/3563). You mentioned node Affinity is immutable in this thread. So I guess even I removed all deprecated labels the upgrade still needs all in-tree PV to be migrate to CSI driver? It seems there is no any other way to modify "node Affinity" other than recreate PV.
